File: Dokie/ViewControllers/Folder/FolderViewModel.m
//
//  FolderViewModel.m
//  Dokie
//
//  Created by Ashamaz Shidov on 28/12/2019.
//  Copyright © 2019 Ashamaz Shidov. All rights reserved.
//

#import "FolderViewModel.h"
#import "Logger.h"
#import "CoreDataInclude.h"
#import "Constants.h"
#import "CoreDataManager.h"
#import "NavigationRouter.h"

@interface FolderViewModel()

@property (nonatomic, strong) NSString *folderName;
@property (nonatomic, strong) CDFolder *folder;
@property (nonatomic, strong) NSArray<NSString *> *subfolders;
@property (nonatomic, strong) NSArray<CDFolder *> *subfolderObjects;
@property (nonatomic, strong) NSArray<NSString *> *documents;
@property (nonatomic, strong) NSArray<CDDocument *> *documentObjects;
@property (nonatomic) BOOL showMenuNavButton;

@property (nonatomic, strong, readonly) NSNotificationCenter *notificationCenter;
@property (nonatomic, strong, readonly) CoreDataManager *coreDataManager;
@property (nonatomic, strong, readonly) NavigationRouter *navigationRouter;

@end

@implementation FolderViewModel

- (instancetype)initWithFolder:(CDFolder *)folder {
    self = [super init];
    if (self) {
        self.folder = folder;
        [self setup];
    }
    return self;
}

- (void)setup {
    @weakify(self);
    [[[[self.notificationCenter rac_addObserverForName:kReloadFolderNotification object:nil]
       takeUntil:self.rac_willDeallocSignal]
      deliverOnMainThread]
     subscribeNext:^(NSNotification *notification) {
        @strongify(self);
        CDFolder *folder = notification.object;
        if ([folder isEqual:self.folder] || (!self.folder && !folder)) {
            [self readFolder];
        }
    }];
    [self readFolder];
    [[[self.coreDataManager folderName:self.folder] deliverOnMainThread] subscribeNext:^(NSString *folderName) {
        @strongify(self);
        self.folderName = folderName;
    }];
    self.showMenuNavButton = self.folder == nil;
}

- (void)addButtonPressed {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:nil
                                                                   message:nil
                                                            preferredStyle:UIAlertControllerStyleActionSheet];
    
    @weakify(self);
    UIAlertAction *addFolder = [UIAlertAction actionWithTitle:NSLocalizedString(@"Add folder", @"Add folder button text in alert view") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        @strongify(self);
        [self createSubfolder];
    }];
    [addFolder setValue:[[UIImage imageNamed:@"folder_small.png"] imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate] forKey:@"image"];
    [alert addAction:addFolder];
    UIAlertAction *addFile = [UIAlertAction actionWithTitle:NSLocalizedString(@"Add document", @"Add document button text in alert view") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        @strongify(self);
        [self createDocument];
    }];
    [addFile setValue:[[UIImage imageNamed:@"file_small.png"] imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate] forKey:@"image"];
    [alert addAction:addFile];
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"Cancel button text in alert view") style:UIAlertActionStyleCancel handler:nil];
    [alert addAction:cancel];
    [self.navigationRouter showAlert:alert];
}

- (void)createSubfolder {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Add folder", @"Add content alert title")
                                                                   message:nil
                                                            preferredStyle:UIAlertControllerStyleAlert];
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = NSLocalizedString(@"Folder name", @"Folder name placeholder");
        textField.autocapitalizationType = UITextAutocapitalizationTypeSentences;
    }];
    
    @weakify(self);
    UIAlertAction *save = [UIAlertAction actionWithTitle:NSLocalizedString(@"Save", @"Save button text in alert view") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        @strongify(self);
        UITextField *folderName = alert.textFields[0];
        [self createSubfolderWithName:folderName.text];
    }];
    [alert addAction:save];
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"Cancel button text in alert view") style:UIAlertActionStyleCancel handler:nil];
    [alert addAction:cancel];
    [self.navigationRouter showAlert:alert];
}

- (void)createSubfolderWithName:(NSString *)subfolderName {
    NSString *loadingGuid = [self.navigationRouter showLoading];
    @weakify(self);
    [[[self.coreDataManager createSubfolderInFolder:self.folder named:subfolderName]
      deliverOnMainThread]
     subscribeError:^(NSError * _Nullable error) {
        @strongify(self);
        [self.navigationRouter hideLoading:loadingGuid];
        [self.navigationRouter showAlertWithTitle:@"" message:NSLocalizedString(@"Can not create folder. Please try again later.", @"Can not create folder error alert.")];
    } completed:^{
        @strongify(self);
        [self.navigationRouter hideLoading:loadingGuid];
        [self readFolder];
    }];
}

- (void)createDocument {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Add document", @"Add content alert title")
                                                                   message:nil
                                                            preferredStyle:UIAlertControllerStyleAlert];
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = NSLocalizedString(@"Document name", @"Document name placeholder");
        textField.autocapitalizationType = UITextAutocapitalizationTypeSentences;
    }];
    
    @weakify(self);
    UIAlertAction *save = [UIAlertAction actionWithTitle:NSLocalizedString(@"Save", @"Save button text in alert view") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        @strongify(self);
        UITextField *docName = alert.textFields[0];
        [self createDocumentWithName:docName.text];
    }];
    [alert addAction:save];
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"Cancel button text in alert view") style:UIAlertActionStyleCancel handler:nil];
    [alert addAction:cancel];
    [self.navigationRouter showAlert:alert];
}

- (void)createDocumentWithName:(NSString *)documentName {
    NSString *loadingGuid = [self.navigationRouter showLoading];
    @weakify(self);
    [[[self.coreDataManager createDocumentInFolder:self.folder named:documentName]
      deliverOnMainThread]
     subscribeNext:^(CDDocument *document) {
        @strongify(self);
        [self.navigationRouter hideLoading:loadingGuid];
        [self.navigationRouter pushDocument:document];
    }
     error:^(NSError * _Nullable error) {
        @strongify(self);
        [self.navigationRouter hideLoading:loadingGuid];
        [self.navigationRouter showAlertWithTitle:@"" message:NSLocalizedString(@"Can not create document. Please try again later.", @"Can not create folder error alert.")];
    }];
}

- (void)longPressedSubfolderAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row < self.subfolders.count) {
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:nil
                                                                       message:nil
                                                                preferredStyle:UIAlertControllerStyleActionSheet];
        
        @weakify(self);
        UIAlertAction *rename = [UIAlertAction actionWithTitle:NSLocalizedString(@"Rename", @"Rename button text in alert view") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            @strongify(self);
            [self renameSubfolderAtIndexPath:indexPath];
        }];
        [alert addAction:rename];
        UIAlertAction *delete = [UIAlertAction actionWithTitle:NSLocalizedString(@"Delete", @"Delete button text in alert view") style:UIAlertActionStyleDestructive handler:^(UIAlertAction * _Nonnull action) {
            @strongify(self);
            [self deleteSubfolderAtIndexPath:indexPath];
        }];
        [alert addAction:delete];
        UIAlertAction *cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"Cancel button text in alert view") style:UIAlertActionStyleCancel handler:nil];
        [alert addAction:cancel];
        [self.navigationRouter showAlert:alert];
    }
}

- (void)longPressedDocumentAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row < self.documents.count) {
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:nil
                                                                       message:nil
                                                                preferredStyle:UIAlertControllerStyleActionSheet];
        
        @weakify(self);
        UIAlertAction *rename = [UIAlertAction actionWithTitle:NSLocalizedString(@"Rename", @"Rename button text in alert view") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            @strongify(self);
            [self renameDocumentAtIndexPath:indexPath];
        }];
        [alert addAction:rename];
        UIAlertAction *delete = [UIAlertAction actionWithTitle:NSLocalizedString(@"Delete", @"Delete button text in alert view") style:UIAlertActionStyleDestructive handler:^(UIAlertAction * _Nonnull action) {
            @strongify(self);
            [self deleteDocumentAtIndexPath:indexPath];
        }];
        [alert addAction:delete];
        UIAlertAction *cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"Cancel button text in alert view") style:UIAlertActionStyleCancel handler:nil];
        [alert addAction:cancel];
        [self.navigationRouter showAlert:alert];
    }
}

- (void)renameSubfolderAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row < self.subfolders.count) {
        NSString *currentName = self.subfolders[indexPath.row];
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Rename folder", @"Rename folder alert title")
                                                                       message:nil
                                                                preferredStyle:UIAlertControllerStyleAlert];
        [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
            textField.placeholder = NSLocalizedString(@"Folder name", @"Folder name placeholder");
            textField.text = currentName;
            textField.autocapitalizationType = UITextAutocapitalizationTypeSentences;
        }];
        
        @weakify(self);
        UIAlertAction *save = [UIAlertAction actionWithTitle:NSLocalizedString(@"Save", @"Save button text in alert view") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            @strongify(self);
            CDFolder *folder = self.subfolderObjects[indexPath.row];
            NSString *loadingGuid = [self.navigationRouter showLoading];
            UITextField *folderName = alert.textFields[0];
            [[[self.coreDataManager renameFolder:folder to:folderName.text]
              deliverOnMainThread]
             subscribeError:^(NSError * _Nullable error) {
                @strongify(self);
                [self.navigationRouter hideLoading:loadingGuid];
                [self.navigationRouter showAlertWithTitle:@"" message:NSLocalizedString(@"Can not rename folder. Please try again later.", @"Can not rename folder error alert.")];
            }
             completed:^{
                @strongify(self);
                [self readFolder];
                [self.navigationRouter hideLoading:loadingGuid];
            }];
        }];
        [alert addAction:save];
        UIAlertAction *cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"Cancel button text in alert view") style:UIAlertActionStyleCancel handler:nil];
        [alert addAction:cancel];
        [self.navigationRouter showAlert:alert];
    }
}

- (void)renameDocumentAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row < self.documents.count) {
        NSString *currentName = self.documents[indexPath.row];
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Rename document", @"Rename document alert title")
                                                                       message:nil
                                                                preferredStyle:UIAlertControllerStyleAlert];
        [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
            textField.placeholder = NSLocalizedString(@"Document name", @"Document name placeholder");
            textField.text = currentName;
            textField.autocapitalizationType = UITextAutocapitalizationTypeSentences;
        }];
        
        @weakify(self);
        UIAlertAction *save = [UIAlertAction actionWithTitle:NSLocalizedString(@"Save", @"Save button text in alert view") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            @strongify(self);
            CDDocument *document = self.documentObjects[indexPath.row];
            NSString *loadingGuid = [self.navigationRouter showLoading];
            UITextField *docName = alert.textFields[0];
            [[[self.coreDataManager renameDocument:document to:docName.text]
              deliverOnMainThread]
             subscribeError:^(NSError * _Nullable error) {
                @strongify(self);
                [self.navigationRouter hideLoading:loadingGuid];
                [self.navigationRouter showAlertWithTitle:@"" message:NSLocalizedString(@"Can not rename folder. Please try again later.", @"Can not rename folder error alert.")];
            }
             completed:^{
                @strongify(self);
                [self readFolder];
                [self.navigationRouter hideLoading:loadingGuid];
            }];
        }];
        [alert addAction:save];
        UIAlertAction *cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"Cancel button text in alert view") style:UIAlertActionStyleCancel handler:nil];
        [alert addAction:cancel];
        [self.navigationRouter showAlert:alert];
    }
}

- (void)deleteSubfolderAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row < self.subfolders.count) {
        NSString *loadingGuid = [self.navigationRouter showLoading];
        @weakify(self);
        [[[self.coreDataManager deleteFolder:self.subfolderObjects[indexPath.row]]
          deliverOnMainThread]
         subscribeError:^(NSError * _Nullable error) {
            @strongify(self);
            [self.navigationRouter hideLoading:loadingGuid];
            [self.navigationRouter showAlertWithTitle:@"" message:NSLocalizedString(@"Some error occured. Please try again later.", @"Try again later text")];
        }
         completed:^{
            @strongify(self);
            [self.navigationRouter hideLoading:loadingGuid];
        }];
    }
}

- (void)deleteDocumentAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row < self.documents.count) {
        NSString *loadingGuid = [self.navigationRouter showLoading];
        @weakify(self);
        [[[self.coreDataManager deleteDocument:self.documentObjects[indexPath.row]]
          deliverOnMainThread]
         subscribeError:^(NSError * _Nullable error) {
            @strongify(self);
            [self.navigationRouter hideLoading:loadingGuid];
            [self.navigationRouter showAlertWithTitle:@"" message:NSLocalizedString(@"Some error occured. Please try again later.", @"Try again later text")];
        }
         completed:^{
            @strongify(self);
            [self.navigationRouter hideLoading:loadingGuid];
        }];
    }
}

- (void)didSelectSubfolderAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row < self.subfolders.count) {
        CDFolder *folder = self.subfolderObjects[indexPath.row];
        [self.navigationRouter pushFolder:folder];
    }
}

- (void)didSelectDocumentAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row < self.documents.count) {
        CDDocument *document = self.documentObjects[indexPath.row];
        [self.navigationRouter pushDocument:document];
    }
}

- (void)readFolder {
    @weakify(self);
    [[[RACSignal combineLatest:@[[self.coreDataManager subfolderNamesInFolder:self.folder],
                                 [self.coreDataManager documentNamesInFolder:self.folder]]]
      deliverOnMainThread]
     subscribeNext:^(RACTuple *tuple) {
        @strongify(self);
        RACTupleUnpack(NSArray<RACTuple *> *folders, NSArray<RACTuple *> *docs) = tuple;
        folders = [folders sortedArrayUsingComparator:^NSComparisonResult(RACTuple *obj1, RACTuple *obj2) {
            NSString *name1 = obj1.second;
            NSString *name2 = obj2.second;
            return [name1 localizedCompare:name2];
        }];
        docs = [docs sortedArrayUsingComparator:^NSComparisonResult(RACTuple *obj1, RACTuple *obj2) {
            NSString *name1 = obj1.second;
            NSString *name2 = obj2.second;
            return [name1 localizedCompare:name2];
        }];
        
        NSMutableArray<NSString *> *subfolders = [NSMutableArray array];
        NSMutableArray<NSString *> *subfolderObjects = [NSMutableArray array];
        for (NSInteger i = 0; i < folders.count; ++i) {
            RACTuple *tuple = folders[i];
            [subfolderObjects addObject:tuple.first];
            [subfolders addObject:tuple.second];
        }
        self.subfolders = [subfolders copy];
        self.subfolderObjects = [subfolderObjects copy];

        NSMutableArray<NSString *> *documents = [NSMutableArray array];
        NSMutableArray<NSString *> *documentObjects = [NSMutableArray array];
        for (NSInteger i = 0; i < docs.count; ++i) {
            RACTuple *tuple = docs[i];
            [documentObjects addObject:tuple.first];
            [documents addObject:tuple.second];
        }
        self.documents = [documents copy];
        self.documentObjects = [documentObjects copy];
    }
     error:^(NSError * _Nullable error) {
        @strongify(self);
        self.documents = nil;
        self.subfolderObjects = nil;
        [self.navigationRouter showAlertWithTitle:@"" message:NSLocalizedString(@"Read error. Please try again later", @"Read error alert text")];
    }];
}

- (void)menuPressed {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:nil
                                                                   message:nil
                                                            preferredStyle:UIAlertControllerStyleActionSheet];
    @weakify(self);
    UIAlertAction *changePassword = [UIAlertAction actionWithTitle:NSLocalizedString(@"Change password", @"Change password button text in alert view") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        @strongify(self);
        [self.navigationRouter showChangePassword];
    }];
    [alert addAction:changePassword];
    UIAlertAction *logout = [UIAlertAction actionWithTitle:NSLocalizedString(@"Logout", @"Logout button text in alert view") style:UIAlertActionStyleDestructive handler:^(UIAlertAction * _Nonnull action) {
        @strongify(self);
        [self.navigationRouter logout];
    }];
    [alert addAction:logout];
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"Cancel button text in alert view") style:UIAlertActionStyleCancel handler:nil];
    [alert addAction:cancel];
    [self.navigationRouter showAlert:alert];
}

#pragma mark -

- (NSNotificationCenter *)notificationCenter {
    return [NSNotificationCenter defaultCenter];
}

- (CoreDataManager *)coreDataManager {
    return [CoreDataManager shared];
}

- (NavigationRouter *)navigationRouter {
    return [NavigationRouter shared];
}

@end

File: Dokie/ViewControllers/Folder/FolderViewModel.swift
//
//  FolderViewModel.swift
//  Dokie
//
//  Created by Ashamaz Shidov on 25.05.2020.
//  Copyright © 2020 Ashamaz Shidov. All rights reserved.
//

import UIKit
import ReactiveCocoa
import ReactiveSwift

class FolderViewModel: NSObject {
    
    let folderName = MutableProperty<String>("")
    let subfolders = MutableProperty<[String]>([])
    let documents = MutableProperty<[String]>([])
    let searchText = MutableProperty<String>("")
    let showMenuNavBarButton: Bool
    let showAddButtonBar = MutableProperty<Bool>(true)
    let noContentText = MutableProperty<String>(String.localized("Folder is empty"))

    private let searchTexts = MutableProperty<[String]>([])

    private let folder: CDFolder?
    private var subfolderObjects = [CDFolder]()
    private var subfolderObjectNames = [String]()
    private var documentObjects = [CDDocument]()
    private var documentObjectNames = [String]()

    private var searchDocumentObjects = [CDDocument]()
    private var folderNameCached = ""
    private let coreDataManager = CoreDataManager.shared
    private let navigationRouter = NavigationRouter.shared
    
    @available(*, unavailable)
    override init() {
        fatalError("Not implemented")
    }
    
    init(folder: CDFolder?) {
        self.folder = folder
        self.showMenuNavBarButton = folder == nil
        super.init()
        setup()
    }
    
    private func setup() {
        NotificationCenter
            .default
            .addObserver(self, selector: #selector(reloadFolder(notification:)), name: NSNotification.Name(rawValue: C.reloadFolderNotification), object: nil)
        
        readFolder()
        
        coreDataManager
            .name(of: folder)
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithResult { [unowned self] result in
                switch result {
                case let .success(name):
                    self.folderName.value = name
                    self.folderNameCached = name
                default:
                    self.folderName.value = ""
                    self.folderNameCached = ""
                }
        }
        
        searchText
            .producer
            .take(during: reactive.lifetime)
            .throttle(1.5, on: QueueScheduler.main)
            .skipRepeats()
            .observe(on: UIScheduler())
            .startWithValues { [unowned self] text in
                if text.isEmpty {
                    self.searchTexts.value = []
                }
                else {
                    self.searchTexts.value = text.components(separatedBy: " ")
                }
        }
        
        searchTexts
            .producer
            .take(during: reactive.lifetime)
            .skipRepeats()
            .observe(on: UIScheduler())
            .startWithValues { [unowned self] texts in
                if texts.isEmpty {
                    self.subfolders.value = self.subfolderObjectNames
                    self.documents.value = self.documentObjectNames
                    self.searchDocumentObjects = []
                    self.showAddButtonBar.value = true
                    self.noContentText.value = String.localized("Folder is empty")
                    self.folderName.value = self.folderNameCached
                }
                else {
                    self.searchForDocumentsTagged(with: texts)
                }
        }
    }
    
    private func searchForDocumentsTagged(with texts: [String]) {
        self.showAddButtonBar.value = false
        self.noContentText.value = String.localized("No documents found")
        self.folderName.value = String.localized("Search...")
        
        let loadingGuid = self.navigationRouter.showLoading()
        
        self.coreDataManager
            .findDocuments(tagedWith: texts)
            .take(during: self.reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithValues { [unowned self] decryptedDocuments in
                let (documentNames, documentObjects) = self.sortDecryptedDocuments(decryptedDocuments)
                self.subfolders.value = []
                self.documents.value = documentNames
                self.searchDocumentObjects = documentObjects

                self.navigationRouter.hideLoading(loadingGuid)
        }
    }
    
    private func sortDecryptedDocuments(_ decryptedDocuments: [CoreDataManager.DecryptedDocumentName]) -> ([String], [CDDocument]) {
        let sortedDocuments = decryptedDocuments.sorted {  (left, right) -> Bool in
            return left.name.localizedCompare(right.name) == .orderedAscending
        }
        
        var documentNames = [String]()
        var documentObjects = [CDDocument]()
        for document in sortedDocuments {
            documentNames.append(document.name)
            documentObjects.append(document.document)
        }

        return (documentNames, documentObjects)
    }
    
    @objc private func reloadFolder(notification: Notification) {
        if let updatedFolder = notification.object as? CDFolder,
            updatedFolder == folder {
            readFolder()
        }
        else if notification.object == nil && folder == nil {
            readFolder()
        }
    }
    
    private func readFolder() {
        coreDataManager
            .subfolderNamesIn(folder: folder)
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithResult { [unowned self] result in
                switch result {
                case let .success(decryptedFolders):
                    let sortedFolders = decryptedFolders.sorted { (left, right) -> Bool in
                        return left.name.localizedCompare(right.name) == .orderedAscending
                    }
                    
                    var subfolderNames = [String]()
                    var subfolderObjects = [CDFolder]()
                    for subfolder in sortedFolders {
                        subfolderNames.append(subfolder.name)
                        subfolderObjects.append(subfolder.folder)
                    }
                    
                    self.subfolderObjectNames = subfolderNames
                    self.subfolderObjects = subfolderObjects

                default:
                    self.subfolderObjectNames = []
                    self.subfolderObjects = []
                    self.navigationRouter.showAlert(title: "", message: "Read error. Please try again later")
                }
                
                if self.searchText.value.isEmpty {
                    self.subfolders.value = self.subfolderObjectNames
                }
        }
        
        coreDataManager
            .documentNamesIn(folder: folder)
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithResult { [unowned self] result in
                switch result {
                case let .success(decryptedDocuments):
                    let (documentNames, documentObjects) = self.sortDecryptedDocuments(decryptedDocuments)
                    self.documentObjectNames = documentNames
                    self.documentObjects = documentObjects
                    
                default:
                    self.documentObjectNames = []
                    self.documentObjects = []
                    self.navigationRouter.showAlert(title: "", message: "Read error. Please try again later")
                }

                if self.searchText.value.isEmpty {
                    self.documents.value = self.documentObjectNames
                }
        }
    }
    
    func addButtonPressed() {
        let alert = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        let addFolder = UIAlertAction(title: String.localized("Add folder"), style: .default) { [unowned self] _ in
            self.createSubfolder()
        }
        addFolder.setValue(UIImage(named: "folder_small.png")?.withRenderingMode(.alwaysTemplate), forKey: "image")
        
        let addFile = UIAlertAction(title: String.localized("Add document"), style: .default) { [unowned self] _ in
            self.createDocument()
        }
        addFile.setValue(UIImage(named: "file_small.png")?.withRenderingMode(.alwaysTemplate), forKey: "image")
        
        let cancel = UIAlertAction(title: String.localized("Cancel"), style: .cancel)
        
        alert.addAction(addFolder)
        alert.addAction(addFile)
        alert.addAction(cancel)
        navigationRouter.showAlert(alert)
    }
    
    private func createSubfolder() {
        let alert = UIAlertController(title: String.localized("Add folder"), message: nil, preferredStyle: .alert)
        alert.addTextField { textField in
            textField.placeholder = String.localized("Folder name")
            textField.autocapitalizationType = .sentences
        }
        
        let save = UIAlertAction(title: String.localized("Save"), style: .default) { [unowned self, alert] _ in
            let textField = alert.textFields![0]
            self.createSubfolder(name: textField.text ?? "")
        }
        
        let cancel = UIAlertAction(title: String.localized("Cancel"), style: .cancel)
        
        alert.addAction(save)
        alert.addAction(cancel)
        navigationRouter.showAlert(alert)
    }
    
    private func createSubfolder(name: String) {
        let loadingGuid = navigationRouter.showLoading()
        coreDataManager
            .createSubfolder(name: name, in: folder)
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithResult { [unowned self] result in
                self.navigationRouter.hideLoading(loadingGuid)
                switch result {
                case .success:
                    self.readFolder()
                case .failure:
                    self.navigationRouter.showAlert(title: "", message: String.localized("Can not create folder. Please try again later."))
                }
        }
    }
    
    private func createDocument() {
        let alert = UIAlertController(title: String.localized("Add document"), message: nil, preferredStyle: .alert)
        alert.addTextField { textField in
            textField.placeholder = String.localized("Document name")
            textField.autocapitalizationType = .sentences
        }
        
        let save = UIAlertAction(title: String.localized("Save"), style: .default) { [unowned self, alert] _ in
            let textField = alert.textFields![0]
            self.createDocument(name: textField.text ?? "")
        }
        
        let cancel = UIAlertAction(title: String.localized("Cancel"), style: .cancel)
        
        alert.addAction(save)
        alert.addAction(cancel)
        navigationRouter.showAlert(alert)
    }
    
    private func createDocument(name: String) {
        let loadingGuid = navigationRouter.showLoading()
        coreDataManager
            .createDocument(name: name, in: folder)
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithResult { [unowned self] result in
                self.navigationRouter.hideLoading(loadingGuid)
                switch result {
                case let .success(document):
                    self.navigationRouter.push(document: document)
                case .failure:
                    self.navigationRouter.showAlert(title: "", message: String.localized("Can not create document. Please try again later."))
                }
        }
    }
    
    func longPressedSubfolder(at indexPath: IndexPath) {
        if indexPath.row < subfolderObjects.count {
            let alert = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
            let rename = UIAlertAction(title: String.localized("Rename"), style: .default) { [unowned self] _ in
                self.renameSubfolder(at: indexPath)
            }
            let delete = UIAlertAction(title: String.localized("Delete"), style: .destructive) { [unowned self] _ in
                self.deleteSubfolder(at: indexPath)
            }
            
            let cancel = UIAlertAction(title: String.localized("Cancel"), style: .cancel)
            
            alert.addAction(rename)
            alert.addAction(delete)
            alert.addAction(cancel)
            navigationRouter.showAlert(alert)
        }
    }
    
    func longPressedDocument(at indexPath: IndexPath) {
        let objects = searchTexts.value.isEmpty ? documentObjects : searchDocumentObjects
        if indexPath.row < objects.count {
            let alert = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
            let rename = UIAlertAction(title: String.localized("Rename"), style: .default) { [unowned self] _ in
                self.renameDocument(at: indexPath)
            }
            let delete = UIAlertAction(title: String.localized("Delete"), style: .destructive) { [unowned self] _ in
                self.deleteDocument(at: indexPath)
            }
            
            let cancel = UIAlertAction(title: String.localized("Cancel"), style: .cancel)
            
            alert.addAction(rename)
            alert.addAction(delete)
            alert.addAction(cancel)
            navigationRouter.showAlert(alert)
        }
    }
    
    func renameSubfolder(at indexPath: IndexPath) {
        if indexPath.row < subfolderObjects.count {
            let currentName = subfolders.value[indexPath.row]
            let alert = UIAlertController(title: String.localized("Rename folder"), message: nil, preferredStyle: .alert)
            alert.addTextField { textField in
                textField.placeholder = String.localized("Folder name")
                textField.text = currentName
                textField.autocapitalizationType = .sentences
            }
            
            let save = UIAlertAction(title: String.localized("Save"), style: .default) { [unowned self, alert] _ in
                let folder = self.subfolderObjects[indexPath.row]
                let loadingGuid = self.navigationRouter.showLoading()
                let folderName = alert.textFields![0].text
                self.coreDataManager
                    .rename(folder: folder, to: folderName ?? "")
                    .take(during: self.reactive.lifetime)
                    .observe(on: UIScheduler())
                    .startWithResult { [unowned self] result in
                        self.navigationRouter.hideLoading(loadingGuid)
                        switch result {
                        case .success:
                            self.readFolder()
                        case .failure:
                            self.navigationRouter.showAlert(title: "", message: String.localized("Can not rename folder. Please try again later."))
                        }
                }
            }
            
            let cancel = UIAlertAction(title: String.localized("Cancel"), style: .cancel)
            
            alert.addAction(save)
            alert.addAction(cancel)
            navigationRouter.showAlert(alert)
        }
    }
    
    func renameDocument(at indexPath: IndexPath) {
        let objects = searchTexts.value.isEmpty ? documentObjects : searchDocumentObjects
        if indexPath.row < objects.count {
            let currentName = documents.value[indexPath.row]
            let alert = UIAlertController(title: String.localized("Rename document"), message: nil, preferredStyle: .alert)
            alert.addTextField { textField in
                textField.placeholder = String.localized("Document name")
                textField.text = currentName
                textField.autocapitalizationType = .sentences
            }
            
            let save = UIAlertAction(title: String.localized("Save"), style: .default) { [unowned self, alert] _ in
                let document = objects[indexPath.row]
                let loadingGuid = self.navigationRouter.showLoading()
                let documentName = alert.textFields![0].text
                self.coreDataManager
                    .rename(document: document, to: documentName ?? "")
                    .take(during: self.reactive.lifetime)
                    .observe(on: UIScheduler())
                    .startWithResult { [unowned self] result in
                        self.navigationRouter.hideLoading(loadingGuid)
                        switch result {
                        case .success:
                            self.readFolder()
                            if !self.searchTexts.value.isEmpty {
                                self.searchForDocumentsTagged(with: self.searchTexts.value)
                            }
                        case .failure:
                            self.navigationRouter.showAlert(title: "", message: String.localized("Can not rename document. Please try again later."))
                        }
                }
            }
            
            let cancel = UIAlertAction(title: String.localized("Cancel"), style: .cancel)
            
            alert.addAction(save)
            alert.addAction(cancel)
            navigationRouter.showAlert(alert)
        }
    }
    
    func deleteSubfolder(at indexPath: IndexPath) {
        if indexPath.row < subfolderObjects.count {
            let loadingGuid = navigationRouter.showLoading()
            coreDataManager
                .delete(folder: subfolderObjects[indexPath.row])
                .take(during: reactive.lifetime)
                .observe(on: UIScheduler())
                .startWithResult { [unowned self] result in
                    self.navigationRouter.hideLoading(loadingGuid)
                    switch result {
                    case .failure:
                        self.navigationRouter.showAlert(title: "", message: String.localized("Some error occured. Please try again later."))
                    default:
                        break
                    }
            }
        }
    }
    
    func deleteDocument(at indexPath: IndexPath) {
        let objects = searchTexts.value.isEmpty ? documentObjects : searchDocumentObjects
        if indexPath.row < objects.count {
            let loadingGuid = navigationRouter.showLoading()
            coreDataManager
                .delete(document: objects[indexPath.row])
                .take(during: reactive.lifetime)
                .observe(on: UIScheduler())
                .startWithResult { [unowned self] result in
                    self.navigationRouter.hideLoading(loadingGuid)
                    switch result {
                    case .failure:
                        self.navigationRouter.showAlert(title: "", message: String.localized("Some error occured. Please try again later."))
                    default:
                        if !self.searchTexts.value.isEmpty {
                            self.searchForDocumentsTagged(with: self.searchTexts.value)
                        }
                        break
                    }
            }
        }
    }
    
    func didSelectSubfolder(at indexPath: IndexPath) {
        if indexPath.row < subfolderObjects.count {
            let folder = subfolderObjects[indexPath.row]
            navigationRouter.push(folder: folder)
        }
    }
    
    func didSelectDocument(at indexPath: IndexPath) {
        let objects = searchTexts.value.isEmpty ? documentObjects : searchDocumentObjects
        if indexPath.row < objects.count {
            let document = objects[indexPath.row]
            navigationRouter.push(document: document)
        }
    }
    
    func menuPressed() {
        let alert = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        let changePassword = UIAlertAction(title: String.localized("Change password"), style: .default) { [unowned self] _ in
            self.navigationRouter.showChangePassword()
        }
        let logout = UIAlertAction(title: String.localized("Logout"), style: .destructive) { [unowned self] _ in
            self.navigationRouter.logout()
        }
        
        let cancel = UIAlertAction(title: String.localized("Cancel"), style: .cancel)
        
        alert.addAction(changePassword)
        alert.addAction(logout)
        alert.addAction(cancel)
        navigationRouter.showAlert(alert)
    }
    
}
